mercadopago/px-ios
# (Added)[PXN-3411] - New service for preparing payments.

## What has changed?

We are adding a new API call just before we call the payment processor to retrieve new discount tokens (and added functionalities in the future). Business motivation, documentation and agreements with Backend can be read [here](https://docs.google.com/document/d/1koIw35Ew91vIYADWPbPeNw7ZrtrYaBUA6EdqvbL91Bo/edit). Endpoint is _/px_mobile/payments/prepare_


Feature will be **On Hold** (but can still receive reviews 🥇) until Backend deploys this new endpoint (ETA next week).

## Description

On this new API call we will ask our backend to retrieve discount tokens, but in future we will extend this functionality for other things (probably tokenization).

### Request/Response body
This API will receive all the info that our backend needs to call _mktools_ (Discounts API):

- Selected payment methods
- Discount configurations
- Preference & charges
- Other needed fields: Marketplace, Public Key, Transaction amount.

And it will return:

- A node with status, that is, a map with the "Key" being a string that represent the request (i.e. "discounts", in future we will have "tokenization" probably) and it's value is another node that contains a status code and a message.
- Requested info for the selected payment methods (i.e. discount tokens).

### Business logic

This API call is **not on the critical path**, so if it fails we continue with the discount tokens we already had (from checkout init call).
Assuming API call was successful, we can have the following cases: 

- Case 1: Selected payment methods don't have discounts. Expected response: Empty response (with no status at all).
- Case 2: Selected payment methods do have discounts. Expected response: Status "OK" and payment methods with their new discount tokens. In this case we have to merge this discount tokens with the checkout init discount info we already had (in future, this will be coming already merged from backend but we trade it off as backend had a lot of work to do).
- Case 3: Selected payment methods do have discounts but there are internal errors on our backend (i.e. the _mktools_ API fails). Expected response: Status "USE_FALLBACK" and payment methods with no discount token. In this case we must use the checkout init discount info we already had.

## How to test:

This feature is being tested extensively together with backend, as it adds overhead on the payment main button and there are many edge cases. Documentation on current tests and how to reproduce can be read [here](https://docs.google.com/spreadsheets/d/1cNXI9n46L6tQHoekGAiYyAaNRHJgO7HMfj05N2l-2FA/edit#gid=147651995).

In any checkout payment (either with default or custom processor) by sniffing the network you should see the POST to the endpoint.

## Prints for layout changes:

None needed.
